torchdata/nodes/io/file_list.py

```diff
@@ -0,0 +1,138 @@
+            "metadata": {
+                "fs_protocol": "s3",        # Filesystem protocol
+                "source": "s3"             # Source type
+            }
+        }
+
+    Examples:
+        >>> # List local files
+        >>> node = FileLister("path/to/local/dir", ["*.txt", "*.json"])
+        >>> for item in node:
+        ...     print(item["data"])  # prints "path/to/local/dir/file1.txt", etc.
+        >>>
+        >>> # List S3 files with anonymous access
+        >>> node = FileLister("s3://my-bucket/prefix", ["*.csv"],
+        ...                   fs_options={"anon": True})
+        >>> for item in node:
+        ...     print(item["data"])  # prints "s3://my-bucket/prefix/file1.csv", etc.
+        >>>
+        >>> # List GCS files
+        >>> node = FileLister("gs://my-bucket/prefix", ["*.parquet"])
+        >>> for item in node:
+        ...     print(item["data"])  # prints "gs://my-bucket/prefix/file1.parquet", etc.
+    """
+
+    SOURCE_KEY = "source"
+    CURRENT_IDX_KEY = "current_idx"
+
+    def __init__(self, uri_base: str, patterns: List[str], fs_options: Optional[Dict[str, Any]] = None):
+        """Initialize the FileLister.
+
+        Args:
+            uri_base: Base URI for file listing (any fsspec-supported URI)
+            patterns: List of glob patterns to match files against
+            fs_options: Options to pass to fsspec filesystem
+        """
+        super().__init__()
+
+        self.uri_base = uri_base
+        self.patterns = patterns if patterns else ["*"]
+        self.fs_options = dict(fs_options or {})
+
+        # Parse protocol from URI using fsspec
+        self.protocol, _ = fsspec.core.split_protocol(uri_base)
+        if self.protocol is None:
+            self.protocol = "file"  # Default to local filesystem
+
+        # Create filesystem instance
+        self.fs = fsspec.filesystem(self.protocol, **self.fs_options)
+
+        # Initialize state
+        self._file_paths = None
+        self._current_idx = 0
+
+    def reset(self, initial_state: Optional[Dict[str, Any]] = None):
+        """Reset node state and populate file paths."""
+        super().reset(initial_state)
+
+        # Always initialize files list
+        self._file_paths = set()
+
+        try:
+            # Use fsspec's glob for file discovery
+            for pattern in self.patterns:
+                pattern_path = os.path.join(self.uri_base, pattern)
+
+                try:
+                    matches = self.fs.glob(pattern_path)
+
+                    # Filter out directories from matches if pattern is **/* or *
+                    if pattern == "**/*" or pattern == "*":
+                        matches = [match for match in matches if self.fs.isfile(match)]
+
+                    self._file_paths.update(matches)
+
+                except Exception as e:
+                    logger.warning(f"Error globbing pattern {pattern}: {e}")
+
+            # Convert to sorted list for deterministic ordering
```

Comment:
What if the file paths or num_files change between runs?
Will reset still work well? 

1. Should we also save num_files in the state to see if num_files match or not?
2. How about file_paths, will that be too expensive?

Reply:
it may depend on how much reproducibility you would like to have in `torchdata`. i like the intermediate idea of checking `num_files` and provide warnings. or - would you like `raise` more?

although strong `file_paths` will ensure the statefulness more strongly, i think it could be too expensive & other nodes like file reader can be supplementary to achieve the goal.

Reply:
@keunwoochoi  thanks.

I think we should raise when num_files do not match.
My reasoning is that if we do not raise an error and there is an increase in the number of files after restart, we might be reading from a different file (after sort) and the user might not even know.

After raising the error, user will know what's happening and can change their filtering logic. 
LMK what you think